go/tasks/pluginmachinery/flytek8s/copilot.go

```diff
@@ -0,0 +1,269 @@
+				inPath = pilot.GetInputPath()
+			}
+
+			c.VolumeMounts = append(c.VolumeMounts, v1.VolumeMount{
+				Name:      cfg.InputVolumeName,
+				MountPath: inPath,
+			})
+		}
+
+		if iFace.Outputs != nil {
+			outPath := cfg.DefaultOutputPath
+			if pilot.GetOutputPath() != "" {
+				outPath = pilot.GetOutputPath()
+			}
+			c.VolumeMounts = append(c.VolumeMounts, v1.VolumeMount{
+				Name:      cfg.OutputVolumeName,
+				MountPath: outPath,
+			})
+		}
+	}
+	return nil
+}
+
+func AddCoPilotToPod(ctx context.Context, cfg config.FlyteCoPilotConfig, coPilotPod *v1.PodSpec, iFace *core.TypedInterface, taskExecMetadata core2.TaskExecutionMetadata, inputPaths io.InputFilePaths, outputPaths io.OutputFilePaths, pilot *core.DataLoadingConfig) error {
+	if pilot == nil || !pilot.Enabled {
+		return nil
+	}
+
+	logger.Infof(ctx, "CoPilot Enabled for task [%s]", taskExecMetadata.GetTaskExecutionID().GetID().TaskId.Name)
+	shareProcessNamespaceEnabled := true
+	coPilotPod.ShareProcessNamespace = &shareProcessNamespaceEnabled
+	if iFace != nil {
+		// TODO think about MountPropagationMode. Maybe we want to use that for acceleration in the future
+		if iFace.Inputs != nil || iFace.Outputs != nil {
+			// This is temporary. we have to mount the flyte data configuration into the pod
+			// TODO Remove the data configuration requirements
+			coPilotPod.Volumes = append(coPilotPod.Volumes, v1.Volume{
+				Name: flyteDataConfigVolume,
+				VolumeSource: v1.VolumeSource{
+					ConfigMap: &v1.ConfigMapVolumeSource{
+						LocalObjectReference: v1.LocalObjectReference{
+							Name: flyteDataConfigMap,
+						},
+					},
+				},
+			})
+		}
+		cfgVMount := v1.VolumeMount{
+			Name:      flyteDataConfigVolume,
+			MountPath: flyteDataConfigPath,
+		}
+
+		if iFace.Inputs != nil {
+			inPath := cfg.DefaultInputDataPath
+			if pilot.GetInputPath() != "" {
+				inPath = pilot.GetInputPath()
+			}
+
+			// TODO we should calculate input volume size based on the size of the inputs which is known ahead of time. We should store that as part of the metadata
+			size := CalculateStorageSize(taskExecMetadata.GetOverrides().GetResources())
+			logger.Infof(ctx, "Adding Input path [%s] of Size [%d] for Task [%s]", size, inPath, taskExecMetadata.GetTaskExecutionID().GetID().TaskId.Name)
+			inputsVolumeMount := v1.VolumeMount{
+				Name:      cfg.InputVolumeName,
+				MountPath: inPath,
+			}
+
+			format := pilot.Format
+			// Lets add the InputsVolume
+			coPilotPod.Volumes = append(coPilotPod.Volumes, DataVolume(cfg.InputVolumeName, size))
+
+			// Lets add the Inputs init container
+			args, err := DownloadCommandArgs(inputPaths.GetInputPath(), outputPaths.GetOutputPrefixPath(), inPath, format, iFace.Inputs)
+			if err != nil {
+				return err
+			}
+			downloader, err := FlyteCoPilotContainer(flyteInitContainerName, cfg, args, inputsVolumeMount, cfgVMount)
+			if err != nil {
+				return err
+			}
+			coPilotPod.InitContainers = append(coPilotPod.InitContainers, downloader)
```

Comment:
Appending will make it the last init container - should we make it the first instead?

Reply:
cannot make the first like in case of IAM wait, but IAM wait is injected. Also, we do not really add the data volume to all init containers, thus not sure, if making it earlier has any uses?